### ISSUE TITLE
Diagnostics: help text for missing left-hand side in arbitrary self type

### ISSUE THREAD

**Issue by yoshuawuyts** [rust-lang/rust#64252]
Forgetting to include the left-hand side of an arbitrary self type can give some confusing errors. This was reported by someone trying to implement a future manually, and the diagnostics threw them off.

I realize diagnostics for arbitrary self types are an open issue, but figured I'd include this variant as to not overlook it. Thanks!

[Link to playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7ade654ee894018a51de5d4d3dcff90d)

## Code
```rust
pub fn hello(Box<Self>) {}
```

## Expected
```txt
error: expected one of `:` or `@`, found `<`
 --> src/lib.rs:1:17
  |
1 | pub fn hello(Box<Self>) {}
  |              ^ expected one of `:` or `@` here
  |
help: if this is a `self` type, give it a parameter name
  |
1 | pub fn hello(self: Box<Self>) {}
  |              ^^^^^^^^^^^^^^^

error: aborting due to previous error
```

## Current
```txt
error: expected one of `:` or `@`, found `<`
 --> src/lib.rs:1:17
  |
1 | pub fn hello(Box<Self>) {}
  |                 ^ expected one of `:` or `@` here

error: aborting due to previous error
```

**Comment by Centril** [comment]
The output with:
```rust
impl Foo for Baz {
    pub fn hello(Box<Self>) {}
}
```
is:
```
error: expected one of `:` or `@`, found `<`
 --> src/lib.rs:2:21
  |
2 |     pub fn hello(Box<Self>) {}
  |                     ^ expected one of `:` or `@` here
  |
  = note: anonymous parameters are removed in the 2018 edition (see RFC 1685)
help: if this was a parameter name, give it a type
  |
2 |     pub fn hello(Box: TypeName<Self>) {}
  |                  ^^^^^^^^^^^^^
help: if this is a type, explicitly ignore the parameter name
  |
2 |     pub fn hello(_: Box<Self>) {}
  |                  ^^^^^^

error: aborting due to previous error
```
